### PR TITLE
Added a crash log flush to the disk after the unexpected crash.

### LINKED ISSUE #51574: system.crash_log is not created sometimes.
Table system.crash_log is not created after a force crash.

**Version**
```bash
./build/programs/clickhouse-client --version
ClickHouse client version 23.6.1.1.
```

**Does it reproduce on recent release?**

Yes.

**How to reproduce**
First, you must delete a table that has already been created, if it exists.
1) Start the ClickHouse server.
2) Drop table system.crash_log 
3) Stop the ClickHouse server.

Now reproduce the problem:
1) Start ClickHouse server.
2) Promt bash command:
```bash
sudo kill -4 $(pgrep clickhouse)
```

If table was successfully created repeat all steps again (drop system.crash_log, restart ClickHouse and kill the process)

**Expected behavior**

Table files are created, and a crash log can be selected by SQL command after server restart: 

```SQL
SELECT * FROM system.crash_log
```

**Actual behavior**
The problem reproduces unstably. Possible variants:
1) No table files are created. (Table data can't be selected.)

```bash
ls  /home/admin/ClickHouse2/ClickHouse/build/programs/data/system/crash_log
ls: cannot access '/home/admin/ClickHouse2/ClickHouse/build/programs/data/system/crash_log': No such file or directory

```
2) A table is created, but SELECT * FROM system.crash_log returns 0 entries:

```bash
tree /home/admin/ClickHouse2/ClickHouse/build/programs/data/system/crash_log
/home/admin/ClickHouse2/ClickHouse/build/programs/data/system/crash_log
├── detached
└── format_version.txt
```

3) Crash_log was successfully created, and table data can be selected (No error here).

### PR DESCRIPTION
**Pull request by Demilivor**
Closes #51574 

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a crash log flush to the disk after the unexpected crash.

### CODE REVIEW

**Comment by Demilivor** [comment]
Please review the changes add the label can be tested.

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 858859d999f36b3adf68a38cd68a2414bd59912d with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51720/858859d999f36b3adf68a38cd68a2414bd59912d/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by hanfei1991 — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Not a bug fix.

File: src/Interpreters/CrashLog.cpp [hunk 1]
```diff
@@ -84,5 +84,6 @@ void collectCrashLog(Int32 signal, UInt64 thread_id, const String & query_id, co
 
         CrashLogElement element{static_cast<time_t>(time / 1000000000), time, signal, thread_id, query_id, trace, trace_full};
         crash_log_owned->add(element);
+        crash_log_owned->flush(true);
```
Comment (alexey-milovidov) [inline src/Interpreters/CrashLog.cpp:87]: This is not the best option, because a crash can happen in multiple threads in a short time.

Comment (alexey-milovidov) [reply src/Interpreters/CrashLog.cpp:87]: Missing a comment.

**Comment by alexey-milovidov** [comment]
I'd better keep it as is, or - as a maximum - lower the flush interval to one second.

**Comment by Demilivor** [comment]
> I'd better keep it as is, or - as a maximum - lower the flush interval to one second.

I will try to test with a reset interval of 1 second.

Although I like the blocking wait more than hoping that the crash log has time to store on disk before the force quit.
Perhaps in the case of multiple crashes, thread should block the wait for just the first crash, instead of waiting for every crash log flush?

**Comment by alexey-milovidov** [comment]
Let's check - what are the mechanics, and why it reproduces?

I like the idea in general, but for the fatal error handler, it makes sense to keep it loose - it is intended to have "best effort" behavior, and adding extra logic there can have side effects, e.g., subsequent errors.

**Comment by Demilivor** [comment]
When a signal is received, the `signalHandler `function from `src/Daemon/BaseDaemon.cpp` starts execution:
1) It writes a signal into signal_pipe.
2) It starts to wait in this loop (`SignalListener `thread should receive this).

```cpp
for (size_t i = 0; i < 300; ++i)
{
/// We will synchronize with the thread printing the messages with an atomic variable to finish earlier.
if (fatal_error_printed.test())
break;

/// signal handler thread wait for 1 second, but it is not enough. (Because flusing logs actually takes more than 1 second (it does file creation, file writing and file closing)
sleepForSeconds(1); 
}
```

3) It calls `call_default_signal_handler`. This function stops everything.

`SignalListener` thread work:
1) It receives a signal from `signalHandler`
2) It creates a new thread running the `SignalListener::onFault `method.

`SignalListener::onFault `thread work: 
1) It adds a crash log entry (`SystemLog<LogElement>::savingThreadFunction()` periodically check entries)
2) It sets `fatal_error_printed `(notifies the signalHandler thread).

`SystemLog<LogElement>::savingThreadFunction()` thread
1) It waits for `flush_interval_milliseconds` (In this case, let it be 1 second.)
2) It starts to write table files.
3) File opening, file writing, and file closing are quite slow operations, those operations do not finish before signalHandler stops work. (When table files already existed on disk, I was not able to find missing crash log entries.)

I hope this diagram may help to understand the problem:
time_diagram

**Comment by alexey-milovidov** [comment]
Thank you! Now everything is clear.

Until recently, the logic was simpler - sleep for 10 seconds and call the default signal handler.
As the default flush interval is 7 seconds, it was typically enough for writing into the crash log.

Then it was changed to waiting and calling the default signal handler as soon as the fatal handling thread finished its work.

Let's think about what we can do without flushing too aggressively...

**Comment by alexey-milovidov** [comment]
The 10-second waiting was initially intended not for the crash_log, but for query_log flushing. It means we have to return it nevertheless.

Let's modify the signal handler, so it will sleep additional 10 seconds after it understood that the fatal handling thread finished its work.

PS. Dealing with it with sleep might sound stupid, but I think it is what is best for this particular code - fatal signal handler...

**Comment by Demilivor** [comment]
10 Seconds should be enough in most of user cases. 

Perhaps It is possible to improve my solution without blocking `SignalListener::onFault` thread:
We can use similar logic from `void SystemLogBase<LogElement>::flush(bool force)` but without waiting for flush finish, using this code:
```cpp
// the code is simplified for readability
template <typename LogElement>
void SystemLogBase<LogElement>::notifyFlush()
{
        std::lock_guard lock(mutex);

        // immediately flush, but do not lock the thread
        requested_flush_up_to = std::max(requested_flush_up_to, this_thread_requested_offset);
        flush_event.notify_all();

    // here we don't wait before flush completes   
}
```

and in `collectCrashLog` we will call `notifyFlush()` instead of `flush(true)`

**Comment by Demilivor** [comment]
Update:
1) Added `SystemLogBase<LogElement>::notifyFlush(bool force) ` public method, it will notify `savingThreadFunction`  thread, but it does not block current thread on waiting before flush finish (I'm not sure about usefulness `force` parameter here).
2) `notifyFlush(true)` is used instead of `flush(true)` in `collectCrashLog`.
3) Added `sleepForSeconds(3)` in `signalHandler` as a timeout for storing crash_log.

**Comment by alexey-milovidov** [comment]
What is happening with the integration tests for PostgreSQL?

**Comment by Demilivor** [comment]
I tried to read these logs: https://s3.amazonaws.com/clickhouse-test-reports/51720/2b77196a470e3181d9812e4a7f5ffd769ab6663e/integration_tests__asan__[6_6].html

But I don't see anything to latch on to.

The failed test_postgresql_replica_database_engine_1 test does not have published logs at all. This is the only failing test (all test cases fail). Other PostgreSQL tests pass.

**Comment by alexey-milovidov** [comment]
The PR is good, and we need to merge it, but we have to address the issue with this test regardless (it cannot be merged without investigation).

**Comment by alexey-milovidov** [comment]
And we still have to figure out what queries in the stress test consume large amount of memory.

**Comment by Demilivor** [comment]
The tests fails seems unsable:
After last update, failed PostgreSQL integration tests are pass, 

Now There another integration tests fail:
```
>       raise Exception("Can't wait Minio to start")
E       Exception: Can't wait Minio to start
```
see:  https://s3.amazonaws.com/clickhouse-test-reports/51720/2c8564e45e216b63f4fb13d30e5dac65b6741758/integration_tests__release__[3_4]/integration_run_parallel0_0.log

**Comment by alexey-milovidov** [comment]
We cannot proceed with failed tests. It is most likely our problem, not yours... but we can hold this PR as a hostage until we will fix all the tests.

**Comment by Demilivor** [comment]
Mergeable check passed:  `257 successful and 1 skipped checks`

**Pull request merged**